transport/mixer/src/channel.rs

```diff
@@ -0,0 +1,356 @@
+
+    /// The mutex lock over the channel failed.
+    #[error("Channel lock failed")]
+    Lock,
+}
+
+/// Receiver object interacting with the mixer channel.
+///
+/// The receiver receives already mixed elements without any knowledge of
+/// the original order.
+pub struct Receiver<T> {
+    channel: Arc<Mutex<Channel<T>>>,
+}
+
+impl<T> Stream for Receiver<T> {
+    type Item = T;
+
+    fn poll_next(self: std::pin::Pin<&mut Self>, cx: &mut std::task::Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut channel = self.channel.lock().unwrap();
+        let now = std::time::SystemTime::now();
+        if channel.sender_count > 0 {
+            if channel.buffer.peek().map(|x| x.0.release_at < now).unwrap_or(false) {
+                let data = channel
+                    .buffer
+                    .pop()
+                    .expect("The value should be present within the same locked access")
+                    .0
+                    .item;
+
+                trace!("yielding a value");
+
+                #[cfg(all(feature = "prometheus", not(test)))]
+                METRIC_QUEUE_SIZE.decrement(1.0f64);
+
+                return Poll::Ready(Some(data));
+            }
+
+            let waker = cx.waker().clone();
+            channel.waker = Some(waker);
+
+            if let Some(next) = channel.buffer.peek() {
+                if let Ok(remaining) = next.0.release_at.duration_since(now) {
+                    trace!("reseting the timer");
+                    channel.timer.reset(remaining);
+                    let timer = std::pin::pin!(&mut channel.timer);
+                    let res = timer.poll(cx);
+
+                    return match res {
+                        Poll::Ready(_) => {
+                            #[cfg(all(feature = "prometheus", not(test)))]
+                            METRIC_QUEUE_SIZE.decrement(1.0f64);
+
+                            Poll::Ready(Some(
+                                channel
+                                    .buffer
+                                    .pop()
+                                    .expect("The value should be present within the locked access")
+                                    .0
+                                    .item,
+                            ))
+                        }
+                        Poll::Pending => Poll::Pending,
+                    };
+                } else {
+                    unreachable!("the previous block would've yielded the value");
+                }
+            }
+
+            Poll::Pending
+        } else {
+            channel.receiver_active = false;
+            Poll::Ready(None)
+        }
+    }
+}
+
+impl<T> Receiver<T> {
+    /// Receive a single delayed mixed item.
+    pub async fn recv(&mut self) -> Option<T> {
+        poll_fn(|cx| self.poll_next_unpin(cx)).await
+    }
+}
+
+/// Instantiate a mixing channel and return the sender and receiver end of the channel.
+pub fn channel<T>() -> (Sender<T>, Receiver<T>) {
```

Comment:
Shouldn't it be called e.g. `mixing_channel` ? So that it is not mistaken with a general purpose channel.

Reply:
`hopr_transport_mixer::channel`.

Reply:
But what if someone imports the namespace